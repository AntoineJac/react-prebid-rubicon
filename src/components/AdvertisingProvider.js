import React, { Component } from 'react';
import Advertising from '../Advertising';
import PropTypes from 'prop-types';
import AdvertisingConfigPropType from './utils/AdvertisingConfigPropType';
import AdvertisingContext from '../AdvertisingContext';

export default class AdvertisingProvider extends Component {
    componentDidMount() {
        if (this.props.active) {
            this.advertising.setup();
        }
    }

    componentDidUpdate() {
        if (this.props.active) {
            this.advertising.setup();
        }
    }

    componentWillUnmount() {
        if (this.advertising) {
            this.advertising.teardown();
        }
    }

    render() {
        this.advertising = new Advertising(this.props.config, this.props.plugins);
        this.activate = this.advertising.activate.bind(this.advertising);
        const data = { activate: this.activate, active: this.props.active };
        return <AdvertisingContext.Provider value={data}>{this.props.children}</AdvertisingContext.Provider>;
    }
}

AdvertisingProvider.propTypes = {
    active: PropTypes.bool,
    config: AdvertisingConfigPropType,
    children: PropTypes.node,
    plugins: PropTypes.arrayOf(
        PropTypes.shape({
            setupPrebid: PropTypes.func,
            setupGpt: PropTypes.func,
            teardownPrebid: PropTypes.func,
            teardownGpt: PropTypes.func
        })
    )
};

AdvertisingProvider.defaultProps = {
    active: true
};
