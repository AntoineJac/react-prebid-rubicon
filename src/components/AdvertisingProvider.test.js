import React from 'react';
import expectSnapshot from '@mt-testutils/expect-snapshot';
import AdvertisingProvider from './AdvertisingProvider';
import { spy } from 'sinon';
import { mount } from 'enzyme';
import config from '../utils/testAdvertisingConfig';

const mockActivate = spy();
const mockSetup = spy();
const mockTeardown = spy();
const mockConstructor = spy();
const mockValueSpy = spy();

jest.mock('../AdvertisingContext', () => ({
    // eslint-disable-next-line react/prop-types
    Provider({ value, children }) {
        mockValueSpy(value);
        return <div id="advertising-context-provider">{children}</div>;
    }
}));

jest.mock(
    '../Advertising',
    () =>
        class {
            constructor(...args) {
                mockConstructor(...args);
            }
            activate(...args) {
                mockActivate(...args);
            }
            setup(...args) {
                mockSetup(...args);
            }
            teardown(...args) {
                mockTeardown(...args);
            }
        }
);

describe('The AdvertisingProvider component', () => {
    it('renders correctly', () =>
        expectSnapshot(
            <AdvertisingProvider config={config}>
                <h1>hello</h1>
            </AdvertisingProvider>
        ));
    describe('when mounted', () => {
        beforeEach(() => mount(<AdvertisingProvider config={config} />));
        it('constructs an Advertising module with the provided configuration', () =>
            void mockConstructor.should.have.been.calledWith(config));
        it('sets up the Advertising module', () => void mockSetup.should.have.been.called);
        it('uses an AdvertisingContext.Provider to pass the activate method of the advertising module', () =>
            expect(mockValueSpy.firstCall.args[0]).toMatchSnapshot());
        afterEach(resetMocks);
    });
    describe('when mounted with active = false', () => {
        beforeEach(() => mount(<AdvertisingProvider config={config} active={false} />));
        it('constructs an Advertising module with the provided configuration', () =>
            void mockSetup.should.not.have.been.called);
        afterEach(resetMocks);
    });
    describe('componentDidUpdate', () => {
        beforeEach(() => {
            const wrapper = mount(<AdvertisingProvider config={config} />);
            wrapper.setProps({ config: config });
            wrapper.update();
        });
        it('constructs an Advertising module with the provided configuration', () =>
            void mockConstructor.should.have.been.calledWith(config));
        it('sets up the Advertising module', () => void mockSetup.should.have.been.called);
        it('uses an AdvertisingContext.Provider to pass the activate method of the advertising module', () =>
            expect(mockValueSpy.firstCall.args[0]).toMatchSnapshot());
        afterEach(resetMocks);
    });
    describe('when mounted with active = false', () => {
        beforeEach(() => {
            const wrapper = mount(<AdvertisingProvider config={config} active={false} />);
            wrapper.setProps({ active: false });
            wrapper.update();
        });
        it('constructs an Advertising module with the provided configuration', () =>
            void mockSetup.should.not.have.been.called);
        afterEach(resetMocks);
    });
    describe('when unmounted', () => {
        let component, componentWillUnmount;
        beforeEach(() => {
            component = mount(<AdvertisingProvider config={config} />);
            component.advertising = true;
            componentWillUnmount = jest.spyOn(component.instance(), 'componentWillUnmount');
            component.unmount();
        });    
        it('unmount Advertising module with the provided configuration', () => {    
            expect(componentWillUnmount).toHaveBeenCalled();
        });
        it('teardown up the Advertising module', () => {
            expect(component.advertising).toBe(true);
        });    
        it('teardown up the Advertising module', () => {
            void mockTeardown.should.have.been.called;
        });      
        afterEach(resetMocks);
    });
});

function resetMocks() {
    mockConstructor.resetHistory();
    mockSetup.resetHistory();
    mockTeardown.resetHistory();
    mockActivate.resetHistory();
    mockValueSpy.resetHistory();
}
