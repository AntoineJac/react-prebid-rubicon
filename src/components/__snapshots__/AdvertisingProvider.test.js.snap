// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The AdvertisingProvider component componentDidUpdate uses an AdvertisingContext.Provider to pass the activate method of the advertising module 1`] = `
Object {
  "activate": [Function],
  "active": true,
}
`;

exports[`The AdvertisingProvider component renders correctly 1`] = `
<div
  id="advertising-context-provider"
>
  <h1>
    hello
  </h1>
</div>
`;

exports[`The AdvertisingProvider component when mounted uses an AdvertisingContext.Provider to pass the activate method of the advertising module 1`] = `
Object {
  "activate": [Function],
  "active": true,
}
`;
